bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java

```diff
@@ -0,0 +1,332 @@
+                .setMessageType(HeaderMessageType.SERIAL_PASSTHRU_REQUEST).setUnknown0((byte) 0x14);
+        sendPacket(step7);
+        byte[] finalResponse = receivePacket();
+        if ((finalResponse[16] & 0xF0) == 0x10) {
+            logger.info("Successful logon to the panel.");
+            isOnline = true;
+        } else {
+            logger.warn("Logon to panel failure.");
+            logoutSequence();
+        }
+        Thread.sleep(300);
+        // TODO check why after a short sleep, a 37 bytes packet is received after logon
+        // ! ! !
+        receivePacket();
+    }
+
+    protected boolean isInialLoginSuccessful(byte[] loginPacketResponse) {
+        byte payloadResponseByte = loginPacketResponse[16];
+
+        byte headerResponseByte = loginPacketResponse[4];
+        switch (headerResponseByte) {
+            case 0x38:
+            case 0x39:
+                if (payloadResponseByte == 0x00) {
+                    logger.info("Login - Login to IP150 - OK");
+                    return true;
+                }
+            case 0x30:
+                logger.warn("Login - Login to IP150 failed - Incorrect password");
+                break;
+            case 0x78:
+            case 0x79:
+                logger.warn("Login - IP module is busy");
+                break;
+        }
+
+        switch (payloadResponseByte) {
+            case 0x01:
+                logger.warn("Login - Invalid password");
+            case 0x02:
+            case 0x04:
+                logger.warn("Login - User already connected");
+            default:
+                logger.warn("Login - Connection refused");
+
+        }
+        return false;
+    }
+
+    @Override
+    public synchronized void logoutSequence() throws IOException {
+        logger.info("Logout packet sent to IP150.");
+        ParadoxIPPacket logoutPacket = new ParadoxIPPacket(LOGOUT_MESAGE_BYTES, true)
+                .setMessageType(HeaderMessageType.SERIAL_PASSTHRU_REQUEST).setUnknown0((byte) 0x14);
+        sendPacket(logoutPacket);
+        close();
+        isOnline = false;
+    }
+
+    protected void sendPacket(ParadoxIPPacket packet) throws IOException {
+        sendPacket(packet.getBytes());
+    }
+
+    private void sendPacket(byte[] packet) throws IOException {
+        ParadoxUtil.printPacket("Tx Packet:", packet);
+        tx.write(packet);
+    }
+
+    protected byte[] receivePacket() throws InterruptedException, IOException {
+        for (int retryCounter = 1; retryCounter <= 3; retryCounter++) {
+            try {
+                byte[] result = new byte[256];
+                rx.read(result);
+                ParadoxUtil.printPacket("RX:", result);
+                if (result[1] > 0 && result[1] + 16 < 256) {
+                    return Arrays.copyOfRange(result, 0, result[1] + 16);
+                }
+            } catch (IOException e) {
+                logger.debug("Unable to retrieve data from RX. {}", e.getMessage());

```

Comment:
Are you never interested in the stacktrace? Would it make sense to not use a placeholder and just pass e